crates/metadata/src/lib.rs

```diff
@@ -73,12 +73,14 @@ pub enum MetadataVersioned {
     /// Version 2 of the contract metadata.
     V2(MetadataVersionDeprecated),
     /// Version 3 of the contract metadata.
-    V3(InkProject),
+    V3(MetadataVersionDeprecated),
+    /// Version 4 of the contract metadata.
+    V4(InkProject),
```

Comment:
That's why the `ink-waterfall` currently fails ‒ `polkadot-js` doesn't yet support this new metadata format, uploading the `.contract` bundles from this PR there is not yet possible, the UI just displays a stack trace and `Uncaught error.` once attempted.

We'll have to create a follow-up issue for `polkadot-js` once this PR is merged.

Reply:
The first question: Do we need a new version of metadata if only the storage layout section is affected by the change?
And the second: I can temporarily return back version 3 to get the size reduction statistic=) Is it okay?

Reply:
I think we need a new storage version. It changes the structure of the metadata. For now it works with the old version because noone used the storage layout, yet. We will try to get all of those changes into the new version: https://github.com/paritytech/ink/issues/1265